### PR TITLE
Manila: add share resource

### PR DESCRIPTION
**Pull request by kayrus**
Part of  #502

### CODE REVIEW

**Comment by kayrus** [comment]
I'm afraid that functional tests won't work on OpenLab. They passed locally on rocky.

**Comment by kayrus** [comment]
As I expected (from the [gophercloud PR](https://github.com/gophercloud/gophercloud/pull/1343#issuecomment-445223898), where I added explicitly skip of manila tests), shares are not created:

```
openstack_sharedfilesystem_share_v2.share_1: Error waiting for share a24fb1e6-1e88-4777-a301-0ed864b12b6b to become available: unexpected state 'error', wanted target 'available'. last error: %!s(<nil>)
```

@jtopjian did you get an update from OpenLab?

**Comment by kayrus** [comment]
Relates to #1345

**Comment by h00130372** [comment]
recheck

**Comment by h00130372** [comment]
recheck

**Comment by h00130372** [comment]
please review https://github.com/theopenlab/openlab-zuul-jobs/pull/398, we decide to set up devstack with manila enabling the LVM driver which operates in driver_handles_share_services=False mode, and please to modify your acctest to create share only with protocol, size and name, like "manila create NFS 1 --name test_share", no need to create network resources anymore.

**Comment by kayrus** [comment]
@h00130372, share creation and extending work, but shrinking fails.

**Comment by kayrus** [comment]
recheck

**Comment by kayrus** [comment]
recheck

**Comment by kayrus** [comment]
recheck

**Comment by kayrus** [comment]
finally
@jtopjian ready for review

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+	}
+
+	sfsClient.Microversion = minManilaShareMicroversion
+
+	isPublic := d.Get("is_public").(bool)
+
+	metadataraw := d.Get("metadata").(map[string]interface{})
+	metadata := make(map[string]string, len(metadataraw))
+	for k, v := range metadataraw {
+		metadata[k] = v.(string)
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's do a type assertion here for extra safety. I've got some changes in queue to do this elsewhere.

```go
for k, v := range metadataRaw {
  if stringVal, ok := v.(string); ok {
    metadata[k] = stringVal
  }
}
```

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+		Metadata:         metadata,
+		ShareNetworkID:   d.Get("share_network_id").(string),
+		AvailabilityZone: d.Get("availability_zone").(string),
+	}
+
+	if v, ok := d.GetOkExists("share_type"); ok {
+		createOpts.ShareType = v.(string)
+	}
+
+	if v, ok := d.GetOkExists("volume_type"); ok {
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: I don't see `volume_type` in the schema.

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+	if err != nil {
+		return fmt.Errorf("Error creating share: %s", err)
+	}
+
+	d.SetId(share.ID)
+
+	// Wait for share to become active before continuing
+	err = waitForSFV2Share(sfsClient, share.ID, "available", []string{"creating", "manage_starting"}, timeout)
+	if err != nil {
+		return err
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's do:

```go
return fmt.Errorf("Error waiting for OpenStack share %s to become available: %s", name, err)
```

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+	if d.HasChange("is_public") {
+		isPublic := d.Get("is_public").(bool)
+		updateOpts.IsPublic = &isPublic
+	}
+
+	if updateOpts != (shares.UpdateOpts{}) {
+		// Wait for share to become active before continuing
+		err = waitForSFV2Share(sfsClient, d.Id(), "available", []string{"creating", "manage_starting", "extending", "shrinking"}, timeout)
+		if err != nil {
+			return err
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's do:

```go
return fmt.Errorf("Error waiting for OpenStack share %s to become available: %s", name, err)
```

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+		})
+
+		if err != nil {
+			return fmt.Errorf("Error updating %s share: %s", d.Id(), err)
+		}
+
+		// Wait for share to become active before continuing
+		err = waitForSFV2Share(sfsClient, d.Id(), "available", []string{"creating", "manage_starting", "extending", "shrinking"}, timeout)
+		if err != nil {
+			return err
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's do:

```go
return fmt.Errorf("Error waiting for OpenStack share %s to become available: %s", name, err)
```

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+		}
+	}
+
+	var pending []string
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's add lines 326-352, 360-364 in a block of:

```go
if d.HasChange("size") {
  ...
}
```

File: openstack/resource_openstack_sharedfilesystem_share_v2.go [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+	}
+
+	if err != nil {
+		return fmt.Errorf("Unable to resize share %s: %s", d.Id(), err)
+	}
+
+	// Wait for share to become active before continuing
+	err = waitForSFV2Share(sfsClient, d.Id(), "available", pending, timeout)
+	if err != nil {
+		return err
```
Comment (jtopjian) [inline openstack/resource_openstack_sharedfilesystem_share_v2.go]: Let's do:

```go
return fmt.Errorf("Error waiting for OpenStack share %s to become available: %s", name, err)
```

File: website/docs/r/sharedfilesystem_share_v2.html.markdown [hunk 8]
```diff
@@ -0,0 +1,112 @@
+---
+layout: "openstack"
+page_title: "OpenStack: sharedfilesystem_share_v2"
+sidebar_current: "docs-openstack-resource-sharedfilesystem-share-v2"
+description: |-
+  Configure a Shared File System share.
+---
+
+# sharedfilesystem\_share\_v2
```
Comment (jtopjian) [inline website/docs/r/sharedfilesystem_share_v2.html.markdown]: `openstack\_sharedfilesystem\_share\_v2`

**Comment by kayrus** [comment]
Added detailed error messages. Now if the share source snapshot doesn't exists, terraform will print 404 detailed error message:

```
Error: Error applying plan:

1 error(s) occurred:

* openstack_sharedfilesystem_share_v2.share_2: 1 error(s) occurred:

* openstack_sharedfilesystem_share_v2.share_2: Error creating share: itemNotFound (404): ShareSnapshotNotFound: Snapshot 70bfbebc-d3ff-4528-8bbb-58422daa280b could not be found.
```

**Comment by jtopjian** [comment]
Implemented as part of #526

**Pull request closed without merging**